autonomous/autonomous.py

```diff
@@ -101,78 +137,69 @@ def drive_to_pick_up(self, state_tm: float, initial_call: bool) -> None:
-        self.goal = path[-1]
+        waypoints: list[Translation2d] = [
+            waypoint if game.is_red() else game.field_flip_translation2d(waypoint)
+            for waypoint in path.waypoints[:-1]
+        ]
+        self.goal = (
+            path.waypoints[-1]
+            if game.is_red()
+            else game.field_flip_translation2d(path.waypoints[-1])
+        )
+        self.goal_heading = (
+            path.final_heading
+            if game.is_red()
+            else game.field_flip_rotation2d(path.final_heading)
+        )
 
         traj_config = TrajectoryConfig(
             maxVelocity=self.MAX_VEL, maxAcceleration=self.MAX_ACCEL
         )
         traj_config.addConstraint(CentripetalAccelerationConstraint(5.0))
-        topRight = Translation2d(self.goal.X() + 0.5, self.goal.Y() + 0.5)
-        bottomLeft = Translation2d(self.goal.X() - 0.5, self.goal.Y() - 0.5)
-        traj_config.addConstraint(
-            RectangularRegionConstraint(
-                bottomLeft, topRight, MaxVelocityConstraint(0.5)
-            )
-        )
-
-        pose = self.chassis.get_pose()
-
-        next_pos = waypoints[0] if waypoints else self.goal.translation()
-        translation = next_pos - pose.translation()
 
         # Generating a trajectory when the robot is very close to the goal is unnecesary, so this
         # return an empty trajectory that starts at the end point so the robot won't move.
-        distance_to_goal = (self.goal.translation() - pose.translation()).norm()
+        distance_to_goal = (self.goal - pose.translation()).norm()
         if distance_to_goal <= 0.01:
             return Trajectory([Trajectory.State(0, 0, 0, pose)])
 
+        next_pos = waypoints[0] if waypoints else self.goal
+        translation = next_pos - pose.translation()
+
         spline_start_momentum_x = translation.x * self.kD
         spline_start_momentum_y = translation.y * self.kD
-
-        goal_spline = Spline3.ControlVector(
-            (self.goal.X(), self.goal.rotation().cos() * self.END_CONTROL_SCALER),
-            (self.goal.Y(), self.goal.rotation().sin() * self.END_CONTROL_SCALER),
-        )
-
         start_point_spline = Spline3.ControlVector(
             (pose.x, spline_start_momentum_x),
             (pose.y, spline_start_momentum_y),
         )
 
-        chassis_velocity = self.chassis.get_velocity()
-        chassis_speed = math.hypot(chassis_velocity.vx, chassis_velocity.vy)
-        traj_config.setStartVelocity(chassis_speed)
+        prev_pos = waypoints[-1] if waypoints else pose.translation()
+        translation = self.goal - prev_pos
+
+        spline_goal_momentum_x = translation.x * self.kD
+        spline_goal_momentum_y = translation.y * self.kD
+        goal_spline = Spline3.ControlVector(
+            (self.goal.X(), spline_goal_momentum_x),
+            (self.goal.Y(), spline_goal_momentum_y),
+        )
+
+        traj_config.setStartVelocity(0.0)
```

Comment:
I am a little concerned about this assumption as we may exit trajectories early if we have picked up a note. Something we should keep an eye on when testing on the real robot.